### DISCUSSION TITLE
winGRASS: using UCRT instead of MSVCRT

### DISCUSSION THREAD

**Discussion by hellik** [OSGeo/grass#2064] [Ideas]
at the moment, building winGRASS is using the MSVCRT within the msys2 build environment:

from [msys2](Environments)

```
MSVCRT (Microsoft Visual C++ Runtime) is available by default on all Microsoft Windows versions, but due to backwards compatibility issues is stuck in the past, not C99 compatible and is missing some features.

  o  It isn't C99 compatible, for example the printf() function family, but...
  o mingw-w64 provides replacement functions to make things C99 compatible in many cases
  o  It doesn't support the UTF-8 locale
  o  Binaries linked with MSVCRT should not be mixed with UCRT ones because the internal structures and data types are different. (More strictly, object files or static libraries built for different targets shouldn't be mixed. DLLs built for different CRTs can be mixed as long as they don't share CRT objects, e.g. FILE*, across DLL boundaries.) Same rule is applied for MSVC compiled binaries because MSVC uses UCRT by default (if not changed).
   o Works out of the box on every Microsoft Windows versions.
```
more and more (open source) software are using the new universal CRT (UCRT - see [Introducing the Universal CRT](https://devblogs.microsoft.com/cppblog/introducing-the-universal-crt/)

from [msys2](Environments)

```
UCRT (Universal C Runtime) is a newer version which is also used by Microsoft Visual Studio by default. It should work and behave as if the code was compiled with MSVC.

    o Better compatibility with MSVC, both at build time and at run time.
    o It only ships by default on Windows 10 and for older versions you have to provide it yourself or depend on the user having it installed.
    o It does support the UTF-8 locale
```
one of the large open source projects, switching to UCRT, is [R](https://blog.r-project.org/2021/12/07/upcoming-changes-in-r-4.2-on-windows/index.html) 

```
R 4.2 for Windows will support UTF-8 as native encoding, which will be a major improvement in encoding support, allowing Windows R users to work with international text and data.

This new feature will require at least Windows 10 (version 1903) on desktop systems, Windows Server 2022 on long-term support server systems or Windows Server 1903 from the semi-annual channel. Older Windows systems will be able to run R, but with the same limitations in the encoding support as in R 4.1 and earlier.

As part of this change, R will require UCRT as the new C runtime for Windows. This means that on desktop systems older than Windows 10 and on server systems older than Windows Server 2016, UCRT will have to be installed before installing R. MSVCRT, the older C runtime, will no longer be supported. R 4.2 will also drop support for 32-bit builds on Windows.
```

this CRT switch in R is enabled via msys2 UCRT (see [Adding UCRT to the build matrix](https://github.com/msys2/MINGW-packages/issues/6901)).

a mid- to long-term suggestion is to switch winGRASS also to use UCRT. adaption in OSGeo4W may be needed.

any thoughts: @landam @jef-n @ninsbl @HuidaeCho

**Comment by ninsbl** [comment]
This sounds very welcome to me. Unfortunately, I have no expertise in this are, just try and error experience.

For the build it might be sufficient to switch the toolchain: (?)
https://packages.msys2.org/package/mingw-w64-ucrt-x86_64-gcc?repo=ucrt64

But the main question is probably what OSGeo4W does. Qoting from your reference above: "Binaries linked with MSVCRT should not be mixed with UCRT ones because the internal structures (...)"

**Reply by HuidaeCho** [reply]
> MSVCRT should not be mixed with UCRT

This... As long as we rely on OSGeo4W, I don't think we can simply switch to UCRT independently.
